### PR TITLE
server: Make float rounding compatible with mysql for types: float, float(m,n) and double(m,n)

### LINKED ISSUE #22791: query result for float type is incompatible with previous version
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

create table tf(f1 float(10,2));
insert into tf values("-34028234.66");
select * from  tf;

### 2. What did you expect to see? (Required)
```
mysql> select * from tf;
+--------------+
| f1           |
+--------------+
| -34028236.00 |
+--------------+
1 row in set (0.00 sec)

```
This result is the same with tidb version v4.0 and mysql.

### 3. What did you see instead (Required)
```
mysql> select * from tf;
+-----------+
| f1        |
+-----------+
| -34028236 |
+-----------+
1 row in set (0.01 sec)

```

Seems the float type logic is changed in the master branch and incompatibility is introduced.

### 4. What is your TiDB version? (Required)
master version with commit hash `9cb9b69b61a76e9cc3f8cfa27f86635c2496cf16`

### PR DESCRIPTION
**Pull request by johan-j**
### What problem does this PR solve?

Issue Number: close #22791

Problem Summary:
- Tidb type float, have to much precision compared to mysql
- Tidb not printing decimals for custom sized float and double types.

### What is changed and how it works?
What's Changed:
This resolve issue #22791: query result for float type is incompatible with previous version

When #21692 was resolved, it created a new problem reported in #22823 (float/double not displaying decimals)
This resolves issues:  #22791, #21692, and a new issue with float I found while working on this (mentioned in comment on issue #22791)

Default width float issue:
```
create table t(a float);
insert into t(a) values( -34028236.66);
select a from t;

mysql8
+-----------+
| a         |
+-----------+
| -34028200 |
+-----------+

tidb:
+-----------+
| a         |
+-----------+
| -34028236 |
+-----------+

```
For this case there was actually even a test case. But the test case expected ` -34028236` as result, which ment it passed even though the value is to precis for mysql.

Previously we relied on go's built in strconv.AppendFloat to both round and format. However after much effort I came up with the solution to treat rounding and formatting separately.
Instead now it works in 2 steps:
1. Call new rounding function that returns in mysql precision (if needed)
2. Use strconv.AppendFloat to format output with correct number of decimals.

This also means that the documentation at: https://docs.pingcap.com/tidb/stable/mysql-compatibility#incompatibility-caused-by-deprecated-features
could mention that deprecated types are now compatible with mysql, even if its encouraged to migrate to Decimal. Many big companies still run older versions like 5.7 of mysql and have no plans to migrate. So keeping it compatible might still be important to get some companies to use TiDB.

The new output formatter supports the below scenarios (which is also the doc/comment in the code):
```
/*
	Formats a float value into a string (byte slice) and returns.

	The rounding/formatting is split into the following cases:
	1. Scientific notation (1.23.e4): Used for big numbers except when value comes from a custom width column or from the select part of the query
	1. mysql.TypeDouble: return string with as much precision as needed (precision = -1)
	2. mysql.TypeFloat: return a rounded value that corresponds to mysql's (precision = 5)
	3. mysql.TypeDouble(m,n):
		3.1 If the columnLength < defaultMySQLPrec64, return string with as much precision as needed (precision = n)
		3.2 else: return rounded string to precision defaultMySQLPrec64, and padded with 0's.
	4. mysql.TypeFloat(m,n): return a rounded value rounded with (precision = 5) and formatted with (precision = n)

	Note: 	Custom float/double column width is deprecated in mysql 8.0.14.
			However for compatibility, we still try to make the output the same as mysql for these types.
*/
```
### Check List

Tests
- Unit test

### Release note
- Resolve issue where float type was not returned in the same way as `mysql8.0`.
- Resolve issue where custom length double/float was returned in incorrect format.

### CODE REVIEW

**Comment by johan-j** [comment]
Closing this PR this since mysql 8.0.14 have deprecated double(m,n) and float(m,n).

**Comment by johan-j** [comment]
I re-open this PR again since I found there are still float formatting issues even for the float type with default column width.

**Comment by johan-j** [comment]
This is still WIP. I have come to the conclusion that we should handle the different cases separately to make each function simpler. Right now we try to create all output formats from same function making quite hard to follow with nested if/else blocks. Next I will try to clean this up to instead have function output format
- float and double (float32 and float64). Show full precision. 
- Scientific output (1.2e3 for example). Show full precision.
- legacy case with float(m,n)/double(m,n) that is deprecated in mysql 8, but still should keep for compatibility. Here we only show the precision defined in <m,n>.

In addition add a function that based on mysql.columnType, colLength, colDecimal decide what is the proper output format function to call.

File: server/util_test.go [hunk 1]
```diff
@@ -253,183 +253,203 @@ func mustDecodeStr(c *C, b []byte) string {
...
 			"-9e-16",
-			-1,
-			64,
-		},
-		{
-			0.11111,
-			"0.111",
-			3,
-			64,
-		},
```
Comment (johan-j) [inline server/util_test.go:315]: I deleted this test because for a column of 3 decimals, the input is already rounded when we store it. So the result is correct in an integration test.
However the result from the function test will not be correct. Let me know if you want me to implement this in the output formatter too, even though its truncated on input.

File: server/util.go [hunk 2]
```diff
@@ -359,52 +353,118 @@ func lengthEncodedIntSize(n uint64) int {
...
+	Formats a float value into a string (byte slice) and returns.
+
+	The rounding/formatting is split into the following cases:
+	1. Scientific notation (1.23.e4): Used for big numbers except when value comes from a custom width column or from the select part of the query
+	1. mysql.TypeDouble: return string with as much precision as needed (precision = -1)
+	2. mysql.TypeFloat: return a rounded value that corresponds to mysql's (precision = 5)
+	3. mysql.TypeDouble(m,n):
+		3.1 If the columnLength < defaultMySQLPrec64, return string with as much precision as needed (precision = n)
+		3.2 else: return rounded string to precision defaultMySQLPrec64, and padded with 0's.
+	4. mysql.TypeFloat(m,n): return a rounded value rounded with (precision = 5) and formatted with (precision = n)
```
[the remark this answers is not in the document]
Comment (johan-j) [reply server/util.go:372]: thank you @djshow832 to both catch the mistake, and also fix it for me!

**Comment by johan-j** [comment]
seems like you guys has no interest to resolve these bugs, so closing this PR instead.

**Comment by AilinKid** [comment]
I'm so sorry for the late response to your PR cause my notification mechanism is broken recently. Even for now, the float problem is still one of the big compatibility problems on our list. 
You can refer some details from: https://github.com/pingcap/tidb/issues/11328

**Comment by johan-j** [comment]
thank you @aierui for showing interest in this pull request. I hope it can resolve the compatibility issues. I put quite some effort into making this PR.

"and if you are interested you can add a more detailed design about your work."
Any suggestion what information is missing? I thought the comment on how the formatter works is already quite detailed. Im happy to clarify where needed! but some specific questions I can answer to make it more clear?

File: server/util.go [hunk 3]
```diff
@@ -359,52 +353,118 @@ func lengthEncodedIntSize(n uint64) int {
...
+	1. Scientific notation (1.23.e4): Used for big numbers except when value comes from a custom width column or from the select part of the query
+	2. mysql.TypeDouble: return string with as much precision as needed (precision = -1)
+	3. mysql.TypeFloat: return a rounded value that corresponds to mysql's (precision = 5)
+	4. mysql.TypeDouble(m,n):
+		4.1 If the columnLength < defaultMySQLPrec64, return string with as much precision as needed (precision = n)
+		4.2 else: return rounded string to precision defaultMySQLPrec64, and padded with 0's.
+	5. mysql.TypeFloat(m,n): return a rounded value rounded with (precision = 5) and formatted with (precision = n)
+
+	Note: 	Custom float/double column width is deprecated in mysql 8.0.14.
+			However for compatibility, we still try to make the output the same as mysql for these types.
```
Comment (AilinKid) [inline server/util.go:375]: better add a mysql referrence link here.

File: server/util.go [hunk 4]
```diff
@@ -359,52 +353,118 @@ func lengthEncodedIntSize(n uint64) int {
...
+*/
+func appendFormatFloat(in []byte, fVal float64, bitSize int, colDecimal uint8, colLength uint32, isFromSelect bool) []byte {
+	isCustomColumnWidth := false
+	prec := types.UnspecifiedLength
+	if colDecimal > 0 && colDecimal != mysql.NotFixedDec {
```
Comment (AilinKid) [inline server/util.go:380]:
```suggestion
	if colDecimal > 0 && int(colDecimal) != mysql.NotFixedDec {
```

File: server/util.go [hunk 5]
```diff
@@ -359,52 +353,118 @@ func lengthEncodedIntSize(n uint64) int {
...
-		if plusPos > 0 {
-			plusPosInOut := len(in) + plusPos
-			out = append(out[:plusPosInOut], out[plusPosInOut+1:]...)
+	}
+
+	switch {
+	case isEFormat:
+		return formatFloatScientificNotation(in, fVal, prec, bitSize)
+	case bitSize == 32 && isCustomColumnWidth:
+		return strconv.AppendFloat(in, fVal, 'f', prec, bitSize)
```
Comment (AilinKid) [inline server/util.go:403]: it's same as the default logic

**Review by AilinKid — COMMENTED** [review]
Rest LGTM, thx for your detailed work.
